main/autogen/autoloader.cc

```diff
@@ -425,11 +422,63 @@ void AutoloadWriter::write(const core::GlobalState &gs, const AutoloaderConfig &
                 string packageName = namespaceName.substr(0, namespaceName.size() - suffixLen);
                 auto pkgSubdir = join(subdir, packageName);
                 FileOps::ensureDir(pkgSubdir);
-                write(gs, alCfg, pkgSubdir, toDelete, *child);
+                populateAutoloadTasksAndCreateDirectories(gs, tasks, alCfg, pkgSubdir, *child);
             } else {
-                write(gs, alCfg, subdir, toDelete, *child);
+                populateAutoloadTasksAndCreateDirectories(gs, tasks, alCfg, subdir, *child);
+            }
+        }
+    }
+}
+}; // namespace
+
+void AutoloadWriter::writeAutoloads(const core::GlobalState &gs, WorkerPool &workers, const AutoloaderConfig &alCfg,
+                                    const std::string &path, const DefTree &root) {
+    vector<RenderAutoloadTask> tasks;
+    populateAutoloadTasksAndCreateDirectories(gs, tasks, alCfg, path, root);
+
+    if (FileOps::exists(path)) {
+        // Clear out files that we do not plan to write.
```

Comment:
This matches the existing logic, but is more explicit. There's room for speedups here.